v2/pkg/manifest/oci-manifest.go

```diff
@@ -251,156 +246,161 @@ func readOperatorCatalog(path string) ([]v1alpha3.DeclarativeConfig, error) {
-		if len(hld) < 4 {
-			return "", fmt.Errorf("versioning of string is not correct %s ", s.Name)
-		}
-		hld[1] = strings.Replace(hld[1], "v", "", -1)
-		end := strings.Split(hld[3], "-")
-		semStr := strings.Join([]string{hld[1], hld[2], end[0]}, ".")
-		version, err := semver.Parse(semStr)
-		if err != nil {
-			return "", err
-		}
+func isInvalidFiltering(pkg v1alpha2.IncludePackage, full bool) bool {
+	return (len(pkg.Channels) > 0 && (pkg.MinVersion != "" || pkg.MaxVersion != "")) ||
+		full && (pkg.MinVersion != "" || pkg.MaxVersion != "")
+}
 
-		if version.Compare(max) == 1 {
-			max = version
-			index = id
+func containsChannel(channels []v1alpha2.IncludeChannel, name string) (bool, int) {
+	for idx, channel := range channels {
+		if channel.Name == name {
+			return true, idx
 		}
 	}
-	return entries[index].Name, nil
+	return false, -1
 }
 
-// semverFindRange - finds the bundles between ranges version
-func semverFindRange(entries []v1alpha3.ChannelEntry, min, max string) ([]string, error) {
-
-	var minVersion semver.Version
-	var maxVersion semver.Version
+func filterBundles(channelEntries []v1alpha3.ChannelEntry, min string, max string, full bool) ([]string, error) {
+	var minVersion, maxVersion semver.Version
 	var err error
-	var results []string
 
-	// parse the min max strings
-	if len(min) > 0 {
-		minVersion, err = semver.Parse(min)
+	if min != "" {
+		minVersion, err = semver.ParseTolerant(min)
 		if err != nil {
-			return []string{}, err
+			return nil, err
 		}
-	} else {
-		minVersion, _ = semver.Parse("0.0.0")
 	}
-	if len(max) > 0 {
-		maxVersion, err = semver.Parse(max)
+
+	if max != "" {
+		maxVersion, err = semver.ParseTolerant(max)
 		if err != nil {
-			return []string{}, err
+			return nil, err
 		}
-	} else {
-		maxVersion, _ = semver.Parse("9.9.9")
 	}
 
-	for _, s := range entries {
-		hld := strings.Split(s.Name, ".")
-		// we are only interested in 1,2,3 positions
-		if len(hld) < 4 {
-			return []string{}, fmt.Errorf("versioning of string is not correct %s ", s.Name)
-		}
-		hld[1] = strings.Replace(hld[1], "v", "", -1)
-		end := strings.Split(hld[3], "-")
-		semStr := strings.Join([]string{hld[1], hld[2], end[0]}, ".")
-		version, err := semver.Parse(semStr)
+	var filtered []string
+	currentHead := semver.MustParse("0.0.0")
+	var currentHeadName string
+
+	for _, chEntry := range channelEntries {
+
+		version, err := getChannelEntrySemVer(chEntry.Name)
 		if err != nil {
-			return []string{}, err
+			return nil, err
 		}
-		if version.Compare(maxVersion) <= 0 && version.Compare(minVersion) >= 1 {
-			results = append(results, s.Name)
+
+		if version.GT(currentHead) || isPreReleaseHead(version, chEntry, currentHeadName) {
+			currentHead = version
+			currentHeadName = chEntry.Name
 		}
+
+		if ((min == "" || version.GTE(minVersion)) && (max == "" || version.LTE(maxVersion))) || isPreReleaseOfFilteredVersion(version, filtered) {
+			if min == "" && max == "" && !full {
```

Comment:
```suggestion
			// In case full ==false and min and max are empty, do not include this bundle:
			// this is the case where there  is no filtering, and where only the channel's head shall be included in the output filter. 
			if min == "" && max == "" && !full {
```